chia/full_node/full_node.py

```diff
@@ -1121,69 +1136,119 @@ async def fetch_block_batches(
+                        # accrue a large state (with no value, since we can validate
+                        # against the CoinStore)
+                        if not extending_main_chain:
+                            if fork_point_height == 0:
+                                fork_info = ForkInfo(-1, -1, self.constants.GENESIS_CHALLENGE)
+                            else:
+                                fork_hash = blockchain.height_to_hash(uint32(fork_point_height - 1))
+                                assert fork_hash is not None
+                                fork_info = ForkInfo(fork_point_height - 1, fork_point_height - 1, fork_hash)
+
+                    blocks_to_validate = await self.filter_blocks(blockchain, blocks, fork_info, vs)
+                    next_validation_state = copy.copy(vs)
+
+                    if len(blocks_to_validate) == 0:
+                        continue
 
-        async def validate_block_batches(
-            inner_batch_queue: asyncio.Queue[Optional[tuple[WSChiaConnection, list[FullBlock]]]]
+                    futures: list[Awaitable[PreValidationResult]] = []
+                    for block in blocks_to_validate:
+                        futures.extend(
+                            await self.prevalidate_blocks(
+                                blockchain,
+                                [block],
+                                peer.peer_info,
+                                vs,
+                                summaries,
+                            )
+                        )
+                    start = time.monotonic()
+                    await output_queue.put((peer, next_validation_state, list(futures), blocks_to_validate))
+                    end = time.monotonic()
+                    if end - start > 1:
+                        self.log.info(f"sync pipeline back-pressure. stalled {end-start:0.2f} seconds on add_block()")
+            except Exception:
+                self.log.exception("Exception validating")
+            finally:
+                # finished signal with None
+                await output_queue.put(None)
+
+        async def ingest_blocks(
+            input_queue: asyncio.Queue[
+                Optional[
+                    tuple[WSChiaConnection, ValidationState, list[Awaitable[PreValidationResult]], list[FullBlock]]
+                ]
+            ],
         ) -> None:
             nonlocal fork_info
             block_rate = 0
             block_rate_time = time.monotonic()
             block_rate_height = -1
             while True:
-                res: Optional[tuple[WSChiaConnection, list[FullBlock]]] = await inner_batch_queue.get()
+                res = await input_queue.get()
                 if res is None:
-                    self.log.debug("done fetching blocks")
+                    self.log.debug("done validating blocks")
                     return None
-                peer, blocks = res
+                peer, vs, futures, blocks = res
```

Comment:
something im missing here, what keeps the order of the blocks we are prevalidating ?